projects/RabbitMQ.Client/client/impl/ModelBase.cs

```diff
@@ -332,38 +334,64 @@ private void HandleCommand(in IncomingCommand cmd)
             }
         }
 
-        protected T ModelRpc<T>(MethodBase method) where T : MethodBase
+        protected void ModelRpc<TMethod>(in TMethod method, ProtocolCommandId returnCommandId)
+            where TMethod : struct, IOutgoingAmqpMethod
         {
             var k = new SimpleBlockingRpcContinuation();
-            var outgoingCommand = new OutgoingCommand(method);
-            MethodBase baseResult;
+            IncomingCommand reply;
             lock (_rpcLock)
             {
-                TransmitAndEnqueue(outgoingCommand, k);
-                baseResult = k.GetReply(ContinuationTimeout).Method;
+                Enqueue(k);
+                Session.Transmit(method);
+                k.GetReply(ContinuationTimeout, out reply);
+            }
+
+            reply.ReturnMethodBuffer();
+
+            if (reply.CommandId != returnCommandId)
+            {
+                throw new UnexpectedMethodException(reply.CommandId, returnCommandId);
+            }
+        }
+
+        protected TReturn ModelRpc<TMethod, TReturn>(in TMethod method, ProtocolCommandId returnCommandId, Func<ReadOnlyMemory<byte>, TReturn> createFunc)
+            where TMethod : struct, IOutgoingAmqpMethod
```

Comment:
You should be able to add a constraint to the TReturn as well as a `struct, IAmqpMethod` right?

Reply:
Correct, but maybe you don't want to return the method itself but rather a subset of properties, this is why I left it unconstrained. (Look at QueuePurge I think, which returns only a subset of it (messages deleted))